includes/abstracts/llms-abstract-controller-user-engagements.php

```diff
@@ -0,0 +1,253 @@
+	protected const TEXT_SYNC_AWARDED_ENGAGEMENT_INSUFFICIENT_PERMISSIONS = 0;
+
+	/**
+	 * @since [version]
+	 *
+	 * @var int
+	 */
+	protected const TEXT_SYNC_AWARDED_ENGAGEMENT_INVALID_TEMPLATE = 1;
+
+	/**
+	 * @since [version]
+	 *
+	 * @var int
+	 */
+	protected const TEXT_SYNC_AWARDED_ENGAGEMENTS_INSUFFICIENT_PERMISSIONS = 2;
+
+	/**
+	 * @since [version]
+	 *
+	 * @var int
+	 */
+	protected const TEXT_SYNC_AWARDED_ENGAGEMENTS_INVALID_NONCE = 3;
+
+	/**
+	 * @since [version]
+	 *
+	 * @var int
+	 */
+	protected const TEXT_SYNC_MISSING_AWARDED_ENGAGEMENT_ID = 4;
+
+	/**
+	 * @since [version]
+	 *
+	 * @var int
+	 */
+	protected const TEXT_SYNC_MISSING_ENGAGEMENT_TEMPLATE_ID = 5;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since [version]
+	 *
+	 * @return void
+	 */
+	public function __construct() {
+
+		add_action( 'init', array( $this, 'maybe_handle_awarded_engagement_sync_actions' ) );
+	}
+
+	/**
+	 * Delete an awarded user engagement.
+	 *
+	 * @since 3.18.0
+	 * @since [version] Permanently delete user engagement via wp_delete_post().
+	 *              Refactored from LLMS_Controller_Achievements::delete() and LLMS_Controller_Certificates::delete().
+	 *
+	 * @param int $post_id WP Post ID of the awarded engagement.
+	 * @return void
+	 */
+	protected function delete( $post_id ) {
+
+		if ( ! is_admin() ) {
+			return;
+		}
+
+		wp_delete_post( $post_id, true );
+	}
+
+	/**
+	 * Returns a translated text of the given type.
+	 *
+	 * @since [version]
+	 *
+	 * @param int   $text_type One of the LLMS_Abstract_Controller_User_Engagements::TEXT_ constants.
+	 * @param array $variables Optional variables that are used in sprintf().
+	 * @return string
+	 */
+	protected function get_text( $text_type, $variables = array() ) {
+
+		return __( 'Invalid text type.', 'lifterlms' );
+	}
+
+	/**
+	 * Handle awarded engagement sync actions.
+	 *
+	 * Errors are added to {@see LLMS_Admin_Metabox::add_error()} to be displayed as an admin notice
+	 * and also returned for unit tests.
+	 *
+	 * If the sync is successful, the {@see llms_redirect_and_exit()} function is called and this method does not return.
+	 *
+	 * @since [version]
+	 *
+	 * @return null|WP_Error
+	 */
+	public function maybe_handle_awarded_engagement_sync_actions() {
+
+		// Validate action.
+		// Invalid actions return a WP_Error for testing purposes and are not displayed to the user.
+		$actions = array(
+			'sync_one'  => "sync_awarded_{$this->engagement_type}",
+			'sync_many' => "sync_awarded_{$this->engagement_type}s",
+		);
+		$action  = llms_filter_input( INPUT_GET, 'action' );
+		if ( ! $action ) {
+			return new WP_Error(
+				"llms-sync-awarded-{$this->engagement_type}s-missing-action",
+				__( 'Sorry, you have not provided any actions.', 'lifterlms' )
+			);
+		} elseif ( ! in_array( $action, $actions, true ) ) {
+			return new WP_Error(
+				"llms-sync-awarded-{$this->engagement_type}s-invalid-action",
+				__( 'You\'re trying to perform an invalid action.', 'lifterlms' )
```

Comment:
```suggestion
				__( "You're trying to perform an invalid action.", 'lifterlms' )
```
Thomas told me once to avoid, where possible, to use _(back)slashes_ in localized strings, to avoid translators messing with them.